### PR TITLE
[SPARK-39843][SQL] Prohibit UDF calls in column DEFAULT expressions

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

Prohibit UDF calls in column DEFAULT expressions.

### Why are the changes needed?

We elect to follow a conservative approach initially with column default values with the understanding that we can expand the feature set later if needed.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New UT coverage.

### CODE REVIEW

**Comment by dtenedor** [comment]
Hi @gengliangwang, this is the change to prohibit UDF expressions for column DEFAULT values in Apache Spark.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala [hunk 1]
```diff
@@ -2504,8 +2504,12 @@ private[sql] object QueryCompilationErrors extends QueryErrorsBase {
   }
 
   def defaultValuesMayNotContainSubQueryExpressions(): Throwable = {
```
Comment (gengliangwang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala:2506]: Shall we change the method call of this method as `defaultValuesMayNotContain("subquery expressions")` for consistency?

**Review by gengliangwang — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala (hunk 1 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/errors/QueryCompilationErrors.scala]: Sure, this cleans up the code. This is done.

**Review by dtenedor — COMMENTED** [review]
Thanks for looking at this!

**Comment by dtenedor** [comment]
Update on this: it turns out that UDF definitions are not present in the separate built-in function analyzer that we currently use for resolving columns DEFAULT expressions. Therefore the new unit test added in this PR returns `"Failed to execute CREATE TABLE command because the destination table column a has a DEFAULT value of udf(42) which fails to resolve as a valid expression: Undefined function: udf. This function is neither a built-in/temporary function, nor a persistent function that is qualified as spark_catalog.default.udf.`

It looks like no extra work is needed to ban UDF calls in DEFAULT expressions for now. I am closing this PR.

**Pull request closed without merging**